services/api/kalvi/api/views/auth.py

```diff
@@ -1,15 +1,114 @@
+        except (ValidationError, PermissionError) as e:
+            return Response({'errors': str(e)}, status=status.HTTP_400_BAD_REQUEST)
 
+#viewset class for changing password, only authenticated users could change password. 
+class UserChangePasswordView(APIView):
+    renderer_classes = [UserRenderer]
+    permission_classes = [IsAuthenticated]
+    def post(self, request, format=None):
+        try:
+            serializer = UserChangePasswordSerialiser(data=request.data)
+            serializer.is_valid(raise_exception=True)
+            user = request.user
+            if not user:
+                return Response({'errors': "Unauthorised Access"}, status=status.HTTP_401_UNAUTHORIZED)
+            user.set_password(serializer.validated_data['password'])
+            user.save()
+            return Response({'msg': 'Password changed successfully'}, status=status.HTTP_200_OK)
+        except ValidationError as e:
+            return Response({'errors': e}, status=status.HTTP_400_BAD_REQUEST)
 
-class SignInEndPoint(APIView):
-    def post(self, request):
-        return Response("SIGNIN ENDPOINT")
+#viewset class for profile view of a user with get method using access tokens.
+class UserProfileView(APIView):
+    renderer_classes = [UserRenderer]
+    permission_classes = [IsAuthenticated]
+    def get(self, request, format=None):
+        try:
+            serializer = UserProfileSerialiser(request.user)
+            return Response(serializer.validated_data, status=status.HTTP_200_OK)
+        except ValidationError as e:
+            return Response({"error": str(e)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+    
+#Viewset class for sending an email to reset the password in case of forgotten passwords.
+class SendPasswordResetEmailView(APIView):
+    renderer_classes = [UserRenderer]
+    def post(self, request, format=None):
+        try:
+            serializer = SendPasswordResetEmailSerializer(data=request.data)
+            serializer.is_valid(raise_exception=True)
+            return Response({'message': 'Password Reset link sent. Please check your Email'}, status=status.HTTP_200_OK)
+        except ValidationError as e:
+            return Response({'error': str(e)}, status=status.HTTP_400_BAD_REQUEST)
+
+#viewset class for changing password from local browser after getting email rest link.
+class UserPasswordResetView(APIView):
+    renderer_classes = [UserRenderer]
+    def post(self, request, uid, token, format=None):
+        try:
+            serializer = UserPasswordResetSerializer(data=request.data, context={'uid':uid, 'token':token})
+            serializer.is_valid(raise_exception=True)
+        except ValidationError as e:
+            return Response({'error': str(e)}, status=status.HTTP_400_BAD_REQUEST)
+        try:
+            id = smart_str(urlsafe_base64_decode(uid))
+            user = User.objects.get(id=id)
+        except (ObjectDoesNotExist, ValueError, TypeError) as e:
```

Comment:
should 404 be used for value error or type error?

Reply:
ohkiee I'll seperate these error

Reply:
i have separated these errors properly